src/dwarf_parser.cpp

```diff
@@ -1,250 +1,213 @@
 #include "dwarf_parser.h"
 
-#ifdef HAVE_LIBDW
+#ifdef HAVE_LIBLLDB
 
 #include "bpftrace.h"
 #include "log.h"
+#include "types.h"
 
-#include <dwarf.h>
-#include <elfutils/libdw.h>
+#include <atomic>
+#include <lldb/API/LLDB.h>
+#include <llvm/Config/llvm-config.h>
+#include <memory>
+#include <string>
 
 namespace bpftrace {
 
-struct FuncInfo {
-  std::string name;
-  Dwarf_Die die;
-};
+std::atomic<size_t> Dwarf::instance_count = 0;
 
-Dwarf::Dwarf(BPFtrace *bpftrace, const std::string &file_path)
-    : bpftrace_(bpftrace), file_path_(file_path)
+Dwarf::Dwarf(BPFtrace *bpftrace, std::string file_path)
+    : bpftrace_(bpftrace), file_path_(std::move(file_path))
 {
-  callbacks.find_debuginfo = dwfl_standard_find_debuginfo;
-  callbacks.section_address = dwfl_offline_section_address;
-  callbacks.debuginfo_path = NULL;
-  dwfl = dwfl_begin(&callbacks);
-  dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1);
-  dwfl_report_end(dwfl, NULL, NULL);
-}
+  if (instance_count++ == 0)
+    lldb::SBDebugger::Initialize();
 
-std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
-                                            const std::string &file_path)
-{
-  std::unique_ptr<Dwarf> dwarf(new Dwarf(bpftrace, file_path));
-  Dwarf_Addr bias;
-  if (dwfl_nextcu(dwarf->dwfl, NULL, &bias) == NULL)
-    return nullptr;
+  debugger_ = lldb::SBDebugger::Create(false);
 
-  return dwarf;
+  lldb::SBError error;
+  target_ = debugger_.CreateTarget(
+      file_path_.c_str(), nullptr, nullptr, false, error);
+  if (!error.Success() || !target_.IsValid()) {
+    throw error;
```

Comment:
I haven't worked out how to trigger this code path, but what happens if we throw `error` while `error.Success()` is true?

Reply:
It will show the following error message to the user:
```
ERROR: Failed to parse DebugInfo: nullptr
```
Not optimal, but I'm not sure in what case we could have no error, nor target. I'll dig a little bit more into it...

Reply:
So we should never reach the second condition in the `if`. This is a quirk of the API that doesn't return a `std::expect` or Rust's `std::Result` kind of object.
Even if we reach this weird edge case, we would get a strange error message, but no crash. Would you be happy to keep it this way?

Reply:
Sounds ok to me then 👍